src/common/remotepermissions.h

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (C) by Olivier Goffart <ogoffart@woboq.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#pragma once
+
+#include <QString>
+#include <QMetaType>
+#include <cstring>
+#include "ocsynclib.h"
+
+namespace OCC {
+
+/**
+ * Class that store in a memory efficient way the remote permission
+ */
+class OCSYNC_EXPORT RemotePermissions
+{
+private:
+    // The first bit tells if the value is set or not
+    // The remaining bits correspond to know if the value is set
+    short _value = 0;
+
+    template <typename Char> // can be 'char' or 'ushort' if conversion from QString
+    void fromArray(const Char *p);
+
+public:
+    enum Permissions {
+        CanWrite = 1,             // W
+        CanDelete = 2,            // D
+        CanRename = 3,            // N
+        CanMove = 4,              // V
+        CanAddFile = 5,           // C
+        CanAddSubDirectories = 6, // K
+        CanReshare = 7,           // R
+        // Note: on the server, this means SharedWithMe, but in discoveryphase.cpp we also set
+        // this permission when the server reports the any "share-types"
+        IsShared = 8,             // S
+        IsMounted = 9,            // M
+        IsMountedSub = 10,        // m (internal: set if the parent dir has IsMounted)
+    };
+    RemotePermissions() = default;
+    RemotePermissions(const char *);
+    RemotePermissions(const QString &);
+
+    QByteArray toString();
+    bool hasPermission(Permissions p) const
+    {
+        if (isNull())
+            return p < IsShared; // when the permissions are unset, everything is allowed
```

Comment:
Mmmh, this kind of logic looks wrong to me in a utility class like this. Maybe separate flag-checking from permission-granting?

Reply:
I don't understand your objection.  It is just setting what flags are the defaults and what are not.

Reply:
Yes, it's the only thing that makes this not a pure storage object. That is what makes it stand out to me. The rest of the class is obvious: it stores a bunch of flags - and here you need to be aware of implicit behavior.

It makes `hasPermission/setPermission` somewhat asymmetric:
```
RemotePermissions p;
assert(p.hasPermission(CanRead));
assert(p.hasPermission(CanWrite));
p.setPermission(CanWrite);
assert(!p.hasPermission(CanRead)); // eh?
```

I expect having this default allows the most convenient use of this object, keep it! Adding explicit documentation would already make me happy.

Reply:
In which case would a null permission field mean "everything is allowed"? In most of the current code it seems to be that `perms.isEmpty()` rather means "unset".

Reply:
The thing is that it ssimplifies the code a lot as i don't have to add the isNull check.  (I don't have to do   (if perm.isNull() || perm.hasPermissions(...)

But if you all say this is so bad design, i can then change that.

Reply:
You could possibly make the default constructor completely equivalent to `RemotePermissions("WDNVCKR")`, also returning `isNull() == true` by setting _value accordingly rather than adding logic in some of the methods.

You currently only call setPermission for stuff `>= IsShared`.